r/subnet - Fix removing ipv6 cidr block from subnet

DrFaust92 opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Closes #10815

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
resource_aws_subnet: allow removing IPV6 CIDR Block
resource_aws_subnet: add plan time validation to `cidr_block`, `ipv6_cidr_block`.
```

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TESTARGS='-run=TestAccAWSSubnet_'
--- PASS: TestAccAWSSubnet_ipv6 (220.53s)
--- PASS: TestAccAWSSubnet_ignoreTags (173.26s)
--- PASS: TestAccAWSSubnet_ipv6 (220.53s)
--- PASS: TestAccAWSSubnet_enableIpv6 (191.56s)
--- PASS: TestAccAWSSubnet_availabilityZoneId (83.43s)
--- PASS: TestAccAWSSubnet_disappears (79.87s)
```

ewbankkit commented:
Verified acceptance tests:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSSubnet_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSSubnet_ -timeout 120m
=== RUN   TestAccAWSSubnet_basic
=== PAUSE TestAccAWSSubnet_basic
=== RUN   TestAccAWSSubnet_ignoreTags
=== PAUSE TestAccAWSSubnet_ignoreTags
=== RUN   TestAccAWSSubnet_ipv6
=== PAUSE TestAccAWSSubnet_ipv6
=== RUN   TestAccAWSSubnet_enableIpv6
=== PAUSE TestAccAWSSubnet_enableIpv6
=== RUN   TestAccAWSSubnet_availabilityZoneId
=== PAUSE TestAccAWSSubnet_availabilityZoneId
=== RUN   TestAccAWSSubnet_disappears
=== PAUSE TestAccAWSSubnet_disappears
=== CONT  TestAccAWSSubnet_basic
=== CONT  TestAccAWSSubnet_availabilityZoneId
=== CONT  TestAccAWSSubnet_disappears
=== CONT  TestAccAWSSubnet_ipv6
=== CONT  TestAccAWSSubnet_enableIpv6
=== CONT  TestAccAWSSubnet_ignoreTags
--- PASS: TestAccAWSSubnet_disappears (34.93s)
--- PASS: TestAccAWSSubnet_availabilityZoneId (38.19s)
--- PASS: TestAccAWSSubnet_basic (38.35s)
--- PASS: TestAccAWSSubnet_ignoreTags (70.50s)
--- PASS: TestAccAWSSubnet_enableIpv6 (86.62s)
--- PASS: TestAccAWSSubnet_ipv6 (94.75s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	94.829s
```

DrFaust92 commented:
rebased and added a few changes:

```
--- PASS: TestAccAWSSubnet_basic (69.39s)
--- PASS: TestAccAWSSubnet_ipv6 (181.27s)
--- PASS: TestAccAWSSubnet_enableIpv6 (160.83s)
--- PASS: TestAccAWSSubnet_disappears (64.57s)
--- PASS: TestAccAWSSubnet_availabilityZoneId (70.12s)
```

ewbankkit commented on aws/resource_aws_subnet.go:
```diff
@@ -273,14 +272,35 @@ func resourceAwsSubnetUpdate(d *schema.ResourceData, meta interface{}) error {
 		// We may need to roll back the state and reattach the old one if this is the case
 
-		_, new := d.GetChange("ipv6_cidr_block")
+		_, newIpv6 := d.GetChange("ipv6_cidr_block")
```
Should just use `d.Get("ipv6_cidr_block").(string)`: #12085.

ewbankkit commented on aws/resource_aws_subnet.go:
```diff
@@ -273,14 +272,35 @@ func resourceAwsSubnetUpdate(d *schema.ResourceData, meta interface{}) error {
 
 		if v, ok := d.GetOk("ipv6_cidr_block_association_id"); ok {
 
+			_, newIpv6AssignOnCreate := d.GetChange("assign_ipv6_address_on_creation")
```
Should just use `d.Get("assign_ipv6_address_on_creation").(bool)`: #12085.

ewbankkit commented on aws/resource_aws_subnet.go:
```diff
@@ -273,14 +272,35 @@ func resourceAwsSubnetUpdate(d *schema.ResourceData, meta interface{}) error {
+				_, err := conn.ModifySubnetAttribute(modifyOpts)
+
+				if err != nil {
+					return err
```
`return fmt.Errorf("error modifying EC2 Subnet (%s) attribute: %w", d.Id(), err)`

ewbankkit commented on aws/resource_aws_subnet.go:
```diff
@@ -292,49 +312,50 @@ func resourceAwsSubnetUpdate(d *schema.ResourceData, meta interface{}) error {
+			if err != nil {
+				//The big question here is, do we want to try to reassociate the old one??
+				//If we have a failure here, then we may be in a situation that we have nothing associated
+				return err
```
`return fmt.Errorf("error associating EC2 Subnet (%s) CIDR block: %w", d.Id(), err)`

ewbankkit commented:
Verified acceptance tests:

```console
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSSubnet_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 2 -run=TestAccAWSSubnet_ -timeout 120m
=== RUN   TestAccAWSSubnet_basic
=== PAUSE TestAccAWSSubnet_basic
=== RUN   TestAccAWSSubnet_ignoreTags
=== PAUSE TestAccAWSSubnet_ignoreTags
=== RUN   TestAccAWSSubnet_ipv6
=== PAUSE TestAccAWSSubnet_ipv6
=== RUN   TestAccAWSSubnet_enableIpv6
=== PAUSE TestAccAWSSubnet_enableIpv6
=== RUN   TestAccAWSSubnet_availabilityZoneId
=== PAUSE TestAccAWSSubnet_availabilityZoneId
=== RUN   TestAccAWSSubnet_disappears
=== PAUSE TestAccAWSSubnet_disappears
=== RUN   TestAccAWSSubnet_outpost
=== PAUSE TestAccAWSSubnet_outpost
=== CONT  TestAccAWSSubnet_basic
=== CONT  TestAccAWSSubnet_availabilityZoneId
--- PASS: TestAccAWSSubnet_basic (25.22s)
=== CONT  TestAccAWSSubnet_outpost
--- PASS: TestAccAWSSubnet_availabilityZoneId (25.94s)
=== CONT  TestAccAWSSubnet_disappears
=== CONT  TestAccAWSSubnet_outpost
    data_source_aws_outposts_outposts_test.go:66: skipping since no Outposts Outpost found
--- SKIP: TestAccAWSSubnet_outpost (1.69s)
=== CONT  TestAccAWSSubnet_ipv6
--- PASS: TestAccAWSSubnet_disappears (22.55s)
=== CONT  TestAccAWSSubnet_enableIpv6
--- PASS: TestAccAWSSubnet_ipv6 (66.58s)
=== CONT  TestAccAWSSubnet_ignoreTags
--- PASS: TestAccAWSSubnet_enableIpv6 (57.08s)
--- PASS: TestAccAWSSubnet_ignoreTags (45.34s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	138.920s
```

ewbankkit commented:
@DrFaust92 Looking good apart from a few minor tweaks.

DrFaust92 commented:
rebased

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -512,10 +544,8 @@ resource "aws_subnet" "test" {
-
-  tags = {
-    Name = %[1]q
+  cidr_block = "10.1.0.0/16"tags = {
```
Looks like there's a little configuration typo here 😅 

```
=== CONT  TestAccAWSSubnet_tags
TestAccAWSSubnet_tags: resource_aws_subnet_test.go:177: Step 3/4 error: terraform failed: exit status 1
stderr:
Error: Missing newline after argument
on config633927574/terraform_plugin_test.tf line 3, in resource "aws_vpc" "test":
3:   cidr_block = "10.1.0.0/16"tags = {
An argument definition must end with a newline.
```

```suggestion
  cidr_block = "10.1.0.0/16"

  tags = {
```

DrFaust92 replied:
thanks, missed this in the rebase

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -477,7 +509,7 @@ resource "aws_vpc" "test" {
 
   tags = {
-    Name = "terraform-testacc-subnet"
+  	Name = "terraform-testacc-subnet"
```
Nit: The two-space formatting previously here and below is expected:

```suggestion
    Name = "terraform-testacc-subnet"
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -616,12 +646,21 @@ resource "aws_subnet" "test" {
 }
 `
 
 const testAccSubnetConfigAvailabilityZoneId = `
+data "aws_availability_zones" "test" {
+  state = "available"
+
+  filter {
+    name   = "opt-in-status"
+    values = ["opt-in-not-required"]
+  }
+}
```
Let's drop in the standard `composeConfig(testAccAvailableAZsNoOptInConfig(), ...)` here 👍

DrFaust92 replied:
done

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -616,12 +648,22 @@ resource "aws_subnet" "test" {
-const testAccSubnetConfigAvailabilityZoneId = `
+func testAccSubnetConfigAvailabilityZoneId() string {
+	return composeConfig(testAccAvailableAZsNoOptInConfig(), `
+data "aws_availability_zones" "test" {
+  state = "available"
+
+  filter {
+    name   = "opt-in-status"
+    values = ["opt-in-not-required"]
+  }
+}
+
```
```suggestion
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -633,13 +675,14 @@ resource "aws_vpc" "test" {
   cidr_block           = "10.1.1.0/24"
   vpc_id               = aws_vpc.test.id
-  availability_zone_id = "usw2-az3"
+  availability_zone_id = data.aws_availability_zones.test.zone_ids[0]
```
```suggestion
  availability_zone_id = data.aws_availability_zones.available.zone_ids[0]
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -515,7 +547,7 @@ resource "aws_vpc" "test" {
 
   tags = {
-    Name = %[1]q
+  	Name = %[1]q
```
```suggestion
    Name = %[1]q
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -537,7 +569,7 @@ resource "aws_vpc" "test" {
 
   tags = {
-    Name = "terraform-testacc-subnet-ipv6"
+  	Name = "terraform-testacc-subnet-ipv6"
```
```suggestion
    Name = "terraform-testacc-subnet-ipv6"
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -547,7 +579,7 @@ resource "aws_subnet" "test" {
 
   tags = {
-    Name = "tf-acc-subnet-ipv6"
+  	Name = "tf-acc-subnet-ipv6"
```
```suggestion
    Name = "tf-acc-subnet-ipv6"
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -558,7 +590,7 @@ resource "aws_vpc" "test" {
 
   tags = {
-    Name = "terraform-testacc-subnet-ipv6"
+  	Name = "terraform-testacc-subnet-ipv6"
```
```suggestion
    Name = "terraform-testacc-subnet-ipv6"
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -570,7 +602,7 @@ resource "aws_subnet" "test" {
 
   tags = {
-    Name = "tf-acc-subnet-ipv6"
+  	Name = "tf-acc-subnet-ipv6"
```
```suggestion
    Name = "tf-acc-subnet-ipv6"
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -581,7 +613,7 @@ resource "aws_vpc" "test" {
 
   tags = {
-    Name = "terraform-testacc-subnet-assign-ipv6-on-creation"
+  	Name = "terraform-testacc-subnet-assign-ipv6-on-creation"
```
```suggestion
    Name = "terraform-testacc-subnet-assign-ipv6-on-creation"
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -593,7 +625,7 @@ resource "aws_subnet" "test" {
 
   tags = {
-    Name = "tf-acc-subnet-assign-ipv6-on-creation"
+  	Name = "tf-acc-subnet-assign-ipv6-on-creation"
```
```suggestion
    Name = "tf-acc-subnet-assign-ipv6-on-creation"
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -604,7 +636,7 @@ resource "aws_vpc" "test" {
 
   tags = {
-    Name = "terraform-testacc-subnet-ipv6-update-cidr"
+  	Name = "terraform-testacc-subnet-ipv6-update-cidr"
```
```suggestion
    Name = "terraform-testacc-subnet-ipv6-update-cidr"
```

bflad commented on aws/resource_aws_subnet_test.go:
```diff
@@ -616,12 +648,22 @@ resource "aws_subnet" "test" {
 
   tags = {
-    Name = "tf-acc-subnet-ipv6-update-cidr"
+  	Name = "tf-acc-subnet-ipv6-update-cidr"
```
```suggestion
    Name = "tf-acc-subnet-ipv6-update-cidr"
```
